### PR TITLE
libkbfs: add putagain

### PR DESCRIPTION
**Pull request by jinyangli**
Augment the bserver interface to include PutAgain

### CODE REVIEW

File: libkbfs/bserver_measured.go [hunk 1]
```diff
@@ -66,6 +66,16 @@ func (b BlockServerMeasured) Put(ctx context.Context, tlfID tlf.ID, id kbfsblock
 	return err
 }
 
+// PutAgain implements the BlockServer interface for BlockServerMeasured.
+func (b BlockServerMeasured) PutAgain(ctx context.Context, tlfID tlf.ID, id kbfsblock.ID,
+	context kbfsblock.Context, buf []byte,
+	serverHalf kbfscrypto.BlockCryptKeyServerHalf) (err error) {
+	b.putTimer.Time(func() {
```
Comment (strib) [inline libkbfs/bserver_measured.go]: We usually have one timer per RPC, so I slightly prefer making a `putAgainTimer` for this.  But if you want to combine the two for a specific reason, I don't feel too strongly against it.

File: libkbfs/bserver_memory.go [hunk 2]
```diff
@@ -175,6 +175,52 @@ func (b *BlockServerMemory) Put(ctx context.Context, tlfID tlf.ID,
 	return refs.put(context, liveBlockRef, "")
 }
 
+// PutAgain implements the BlockServer interface for BlockServerMemory.
+func (b *BlockServerMemory) PutAgain(ctx context.Context, tlfID tlf.ID,
+	id kbfsblock.ID, context kbfsblock.Context, buf []byte,
+	serverHalf kbfscrypto.BlockCryptKeyServerHalf) (err error) {
+
```
Comment (strib) [inline libkbfs/bserver_memory.go]: Extra blank line.

File: libkbfs/bserver_memory.go [hunk 3]
```diff
@@ -175,6 +175,52 @@ func (b *BlockServerMemory) Put(ctx context.Context, tlfID tlf.ID,
...
+	if err := checkContext(ctx); err != nil {
+		return err
+	}
+
+	defer func() {
+		err = translateToBlockServerError(err)
+	}()
+	b.log.CDebugf(ctx, "BlockServerMemory.PutAgain id=%s tlfID=%s context=%s "+
+		"size=%d", id, tlfID, context, len(buf))
+	b.lock.Lock()
```
Comment (strib) [inline libkbfs/bserver_memory.go]: Everything from this point on looks copied from `BlockServerMemory.Put()` (except this is missing the key half check).  Can we refactor it into a shared helper function?

File: libkbfs/bserver_memory.go [hunk 4]
```diff
@@ -175,6 +175,52 @@ func (b *BlockServerMemory) Put(ctx context.Context, tlfID tlf.ID,
...
+	serverHalf kbfscrypto.BlockCryptKeyServerHalf) (err error) {
+
+	if err := checkContext(ctx); err != nil {
+		return err
+	}
+
+	defer func() {
+		err = translateToBlockServerError(err)
+	}()
+	b.log.CDebugf(ctx, "BlockServerMemory.PutAgain id=%s tlfID=%s context=%s "+
```
Comment (strib) [inline libkbfs/bserver_memory.go]: Should we be calling `validateBlockPut` here like we do in `Put()`?

Comment (strib) [inline libkbfs/bserver_memory.go]: (Though I'm not sure why `Put()` allows for adding references -- seems like it already implemented `PutAgain`!)

File: libkbfs/interfaces.go [hunk 5]
```diff
@@ -1398,6 +1398,11 @@ type BlockServer interface {
 	Put(ctx context.Context, tlfID tlf.ID, id kbfsblock.ID, context kbfsblock.Context,
 		buf []byte, serverHalf kbfscrypto.BlockCryptKeyServerHalf) error
 
+	// PutAgain re-stores a previously deleted block under the same ID
+	// with the same data
```
Comment (strib) [inline libkbfs/interfaces.go]: Please end with punctuation.

File: libkbfs/test_common.go [hunk 6]
```diff
@@ -79,9 +79,10 @@ func MakeTestBlockServerOrBust(t logger.TestLogBackend,
 	}
 }
 
-// MakeTestConfigOrBust creates and returns a config suitable for
-// unit-testing with the given list of users.
-func MakeTestConfigOrBust(t logger.TestLogBackend,
+// MakeTestConfigOrBustLoggedIn creates and returns a config suitable for
+// unit-testing with the given list of users. loggedInIndex specifies the
+// index (in the list) of the user being logged in
+func MakeTestConfigOrBustLoggedIn(t logger.TestLogBackend, loggedInIndex int,
```
Comment (strib) [inline libkbfs/test_common.go]: Are you using these changes somewhere else?  The change is fine, I'm just curious why it's in this PR.

**Comment by strib** [comment]
@jinyangli: I rebased this one too, to pull in the existing vendoring updates.  Please fetch/reset to this one before you work on it next.

File: libkbfs/test_common.go (hunk 6 above)
Comment (jinyangli) [inline libkbfs/test_common.go]: This is to used to clean up the tests in kbfs-server/bserver. The cleaning of tests is orthogonal to putAgain, but the old tests no longer compiled after I vendored in new kbfs changes and I thought it's a good time to clean up those tests...

**Pull request merged**